pkg/lib/cockpit-fsinfo.ts

```diff
@@ -0,0 +1,137 @@
+"use strict";
+
+function is_json_dict(value: cockpit.JsonValue): value is cockpit.JsonObject {
+    return value?.constructor === Object;
+}
+
+/* RFC 7396 — JSON Merge Patch — functional */
+function json_merge(current: cockpit.JsonValue, patch: cockpit.JsonValue): cockpit.JsonValue {
+    if (is_json_dict(patch)) {
+        const updated = is_json_dict(current) ? { ...current } : { };
+
+        for (const [key, value] of Object.entries(patch)) {
+            if (value === null) {
+                delete updated[key];
+            } else {
+                updated[key] = json_merge(updated[key], value);
+            }
+        }
+
+        return updated;
+    } else {
+        return patch;
+    }
+}
+
+interface FileInfo {
+    type?: string;
+    tag?: string;
+    mode?: number;
+    size?: number;
+    uid?: number;
+    user?: string | number;
+    gid?: number;
+    group?: string | number;
+    mtime?: number;
+    content?: string;
+    target?: string;
+    entries?: {
+        [filename: string]: FileInfo;
+    };
+    targets?: {
+        [filename: string]: FileInfo;
+    };
+}
+
+interface FsInfoError {
+    problem?: string;
+    message?: string;
+    errno?: string;
+}
+
+interface FileInfoState {
+    info: FileInfo | null;
+    error: FsInfoError | null;
+}
+
+interface FsInfoHandle {
+    close(): void;
+    effect(callback: ((state: FileInfoState) => void)): void;
+    entry(name: string): FileInfo | null;
+    state: FileInfoState;
+    target(name: string): FileInfo | null;
+}
+
+export function fsinfo(path: string, attrs: string[], options?: cockpit.JsonObject) {
+    const self: FsInfoHandle = {
+        close,
+        effect,
+        entry,
+        state: {
+            info: null,
+            error: null,
+        },
+        target,
+    };
+
+    const callbacks: ((state: FileInfoState) => void)[] = [];
+
+    function close() {
+        channel.close();
+    }
+
+    function effect(callback: (state: FileInfoState) => void) {
```

Comment:
I won't fight *very* hard for this, but at least a little: Let's *please* not call this `effect()`, it's super confusing as an API. The standard JS approach is to make this an [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) and attach a `changed` or similar event handler to it.

Reply:
My 2¢:
  - I hate the event mixin API
  - we have a lot of other areas in cockpit.js which implement their own callback list mechanism.  We even have helper functions for it.
  - The name `effect()` is definitely suspect, as is the very react-flavoured API (returning the unregister callable).
  - Maybe we should change it anyway, just for uniformity with everything else in `cockpit.js`...

Reply:
> we have a lot of other areas in cockpit.js which implement their own callback list mechanism

Yeah, but I wouldn't exactly call cockpit.js a role model for good/modern JS. It is really bad from today's POV, it was created pre-ES5 even, with jquery.

Reply:
So this seems to be the only real blocker? I am happy to rewrite `.effect` into something else the question is.. what is a Modern JS API :)

Reply:
But should we then just move this into cockpit? The issue is kinda how we then expose fsinfo, `cockpit.file.info`?